internal/datastore/common/relationships.go

```diff
@@ -0,0 +1,181 @@
+	colsToSelect = append(colsToSelect, field)
+	return colsToSelect
+}
+
+// Querier is an interface for querying the database.
+type Querier[R Rows] interface {
+	QueryFunc(ctx context.Context, f func(context.Context, R) error, sql string, args ...any) error
+}
+
+// Rows is a common interface for database rows reading.
+type Rows interface {
+	Scan(dest ...any) error
+	Next() bool
+	Err() error
+}
+
+type closeRowsWithError interface {
+	Rows
+	Close() error
+}
+
+type closeRows interface {
+	Rows
+	Close()
+}
+
+// QueryRelationships queries relationships for the given query and transaction.
+func QueryRelationships[R Rows, C ~map[string]any](ctx context.Context, queryInfo QueryInfo, sqlStatement string, args []any, span trace.Span, tx Querier[R], withIntegrity bool) (datastore.RelationshipIterator, error) {
+	defer span.End()
+
+	colsToSelect := make([]any, 0, 8)
+	var resourceObjectType string
+	var resourceObjectID string
+	var resourceRelation string
+	var subjectObjectType string
+	var subjectObjectID string
+	var subjectRelation string
+	var caveatName sql.NullString
+	var caveatCtx C
+	var expiration *time.Time
+
+	var integrityKeyID string
+	var integrityHash []byte
+	var timestamp time.Time
+
+	colsToSelect = StaticValueOrAddColumnForSelect(colsToSelect, queryInfo, queryInfo.Schema.ColNamespace, &resourceObjectType)
+	colsToSelect = StaticValueOrAddColumnForSelect(colsToSelect, queryInfo, queryInfo.Schema.ColObjectID, &resourceObjectID)
+	colsToSelect = StaticValueOrAddColumnForSelect(colsToSelect, queryInfo, queryInfo.Schema.ColRelation, &resourceRelation)
+	colsToSelect = StaticValueOrAddColumnForSelect(colsToSelect, queryInfo, queryInfo.Schema.ColUsersetNamespace, &subjectObjectType)
+	colsToSelect = StaticValueOrAddColumnForSelect(colsToSelect, queryInfo, queryInfo.Schema.ColUsersetObjectID, &subjectObjectID)
+	colsToSelect = StaticValueOrAddColumnForSelect(colsToSelect, queryInfo, queryInfo.Schema.ColUsersetRelation, &subjectRelation)
+
+	if !queryInfo.SkipCaveats || queryInfo.Schema.ColumnOptimization == ColumnOptimizationOptionNone {
+		colsToSelect = append(colsToSelect, &caveatName, &caveatCtx)
+	}
+
+	colsToSelect = append(colsToSelect, &expiration)
+
+	if withIntegrity {
+		colsToSelect = append(colsToSelect, &integrityKeyID, &integrityHash, &timestamp)
+	}
+
+	if len(colsToSelect) == 0 {
+		var unused int
+		colsToSelect = append(colsToSelect, &unused)
+	}
+
+	return func(yield func(tuple.Relationship, error) bool) {
+		err := tx.QueryFunc(ctx, func(ctx context.Context, rows R) error {
+			var r Rows = rows
+			if crwe, ok := r.(closeRowsWithError); ok {
+				defer LogOnError(ctx, crwe.Close)
+			} else if cr, ok := r.(closeRows); ok {
+				defer cr.Close()
+			}
+
+			span.AddEvent("Query issued to database")
+			relCount := 0
+			for rows.Next() {
+				if err := rows.Scan(colsToSelect...); err != nil {
+					return fmt.Errorf(errUnableToQueryRels, fmt.Errorf("scan err: %w", err))
```

Comment:
shouldn't this error be yielded?

Reply:
It is, by the yield outside of the QueryFunc call